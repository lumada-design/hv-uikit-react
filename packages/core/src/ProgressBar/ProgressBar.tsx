import { useDefaultProps } from "../hooks/useDefaultProps";

import { HvBaseProps } from "../types/generic";
import { HvTypography, HvTypographyProps } from "../Typography";

import { clamp } from "../utils/helpers";
import { ExtractNames } from "../utils/classes";

import { staticClasses, useClasses } from "./ProgressBar.styles";

export { staticClasses as progressBarClasses };

export type HvProgressBarClasses = ExtractNames<typeof useClasses>;

export type HvProgressBarStatus = "inProgress" | "completed" | "error";

/**
 * ProgressBar provides feedback about a process that is taking place in the application.
 */
export interface HvProgressBarProps extends HvBaseProps {
  /** The value of the progress bar. */
  value: number;
  /**
   * The status of the progress bar.
   *
   * inProgress is black, error is red and completed is green.
   *
   * When uncontrolled and unspecified it will default to "inProgress".
   */
  status?: HvProgressBarStatus;
  /** Aria Properties passed on to the progress bar. */
  labelProps?: HvTypographyProps;
  /** A Jss Object used to override or extend the styles applied to the component. */
  classes?: HvProgressBarClasses;
  /**
   * True if the text label should not be displayed.
   */
  hideLabel?: boolean;
}

export const HvProgressBar = (props: HvProgressBarProps) => {
  const {
    className,
    classes: classesProp,
    value = 0,
    status = "inProgress",
    labelProps,
    hideLabel,
    ...others
  } = useDefaultProps("HvProgressBar", props);

  const { classes, cx } = useClasses(classesProp);

  const clampedValue = clamp(value, 100);

  return (
    <div
      className={cx(classes.root, classes.progress, className)}
      role="progressbar"
      aria-valuemin={0}
      aria-valuemax={100}
      aria-valuenow={clampedValue}
      {...others}
    >
      <div className={classes.progressContainer}>
        {!hideLabel && (
          <HvTypography
            className={classes.progressBarLabel}
            variant="caption2"
            style={{ width: `${clampedValue}%` }}
            {...labelProps}
          >
            {`${clampedValue}%`}
          </HvTypography>
        )}
        <div className={classes.progressBarContainer}>
          <div
            style={{ width: `${clampedValue}%` }}
            className={cx(classes.progressBar, {
              [classes.progressDone]: status === "completed",
              [classes.progressError]: status === "error",
            })}
          />
        </div>
      </div>
    </div>
  );
};
