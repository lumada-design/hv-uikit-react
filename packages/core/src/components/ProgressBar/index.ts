export { default as progressBarClasses } from "./progressBarClasses";
export * from "./progressBarClasses";
export * from "./ProgressBar";
