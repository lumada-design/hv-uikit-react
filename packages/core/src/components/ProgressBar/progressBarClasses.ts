import { getClasses } from "utils";

export type HvProgressBarClasses = {
  root?: string;
  progress?: string;
  progressBar?: string;
  progressBarLabel?: string;
};

const classKeys: string[] = [
  "root",
  "progress",
  "progressBar",
  "progressBarLabel",
];

const progressBarClasses = getClasses<HvProgressBarClasses>(
  classKeys,
  "HvProgressBar"
);

export default progressBarClasses;
