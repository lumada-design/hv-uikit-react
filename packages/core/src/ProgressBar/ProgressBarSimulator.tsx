import { AriaAttributes, useEffect, useState } from "react";

import { HvButton } from "../Button";

import { HvProgressBar, HvProgressBarStatus } from "./ProgressBar";

export const ProgressBarSimulator = ({
  inc,
  error,
  ariaLabel,
}: {
  inc: (v: number) => number;
  error?: number;
  ariaLabel?: AriaAttributes["aria-label"];
  ariaLive?: AriaAttributes["aria-live"];
}) => {
  const [status, setStatus] = useState("inProgress");
  const [value, setValue] = useState(0);
  const [run, setRun] = useState(false);

  useEffect(() => {
    const interval = setInterval(() => {
      if (value === 100) {
        setRun(false);
        clearInterval(interval);
        setStatus("completed");
      } else if (error && value === error) {
        clearInterval(interval);
        setStatus("error");
      } else if (run) setValue(inc);
    }, 150);
    return () => clearInterval(interval);
  }, [inc, value, run, error]);

  const reset = () => {
    setValue(0);
    setStatus("inProgress");
  };

  const start = () => {
    setRun(true);
  };

  return (
    <div style={{ width: 400 }}>
      <HvProgressBar
        value={value}
        status={status as HvProgressBarStatus}
        aria-label={ariaLabel}
        aria-busy={status === "inProgress"}
      />
      <div style={{ marginTop: 10 }}>
        <HvButton onClick={start}>Start</HvButton>
        <HvButton style={{ marginLeft: 10 }} onClick={reset}>
          Reset
        </HvButton>
      </div>
    </div>
  );
};
