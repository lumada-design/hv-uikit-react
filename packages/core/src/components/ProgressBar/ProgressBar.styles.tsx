import styled from "@emotion/styled";
import { theme } from "@hitachivantara/uikit-styles";
import { HvTypography } from "components";
import { transientOptions } from "utils/transientOptions";
import { HvProgressBarStatus } from "./ProgressBar";

export const StyledRoot = styled("div")({
  display: "flex",
  flexDirection: "column",
  alignItems: "center",
  justifyContent: "center",

  // progress
  width: "100%",
});

export const StyledProgressContainer = styled("div")({
  width: "100%",
});

export const StyledValue = styled(HvTypography)({
  backgroundColor: "transparent",
  display: "flex",
  justifyContent: "end",
});

export const StyledProgressBarContainer = styled("div")({
  display: "flex",
  width: "100%",
  height: 4,
  backgroundColor: theme.colors.atmo4,
});

export const StyledProgressBar = styled(
  "div",
  transientOptions
)(({ $status }: { $status?: HvProgressBarStatus }) => ({
  backgroundColor: "#000",
  height: 4,
  ...($status === "completed" && {
    backgroundColor: theme.colors.sema1,
  }),
  ...($status === "error" && {
    backgroundColor: theme.colors.sema4,
  }),
}));
