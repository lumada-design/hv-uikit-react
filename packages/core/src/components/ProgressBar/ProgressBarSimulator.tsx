import { useEffect, useState } from "react";
import { HvButton } from "components";
import { HvProgressBar, HvProgressBarStatus } from "./ProgressBar";

export const ProgressBarSimulator = ({
  inc,
  error,
  ariaLabel,
  ariaLive,
}: {
  inc: (v: number) => number;
  error?: number;
  ariaLabel?: string;
  ariaLive?: string;
}) => {
  const [status, setStatus] = useState("inProgress");
  const [value, setValue] = useState(0);
  const [run, setRun] = useState(false);

  useEffect(() => {
    const interval = setInterval(() => {
      if (value === 100) {
        setRun(false);
        clearInterval(interval);
        setStatus("completed");
      } else if (error && value === error) {
        clearInterval(interval);
        setStatus("error");
      } else if (run) setValue(inc);
    }, 150);
    return () => clearInterval(interval);
  }, [inc, value, run, error]);

  const reset = () => {
    setValue(0);
    setStatus("inProgress");
  };

  const start = () => {
    setRun(true);
  };

  return (
    <div style={{ width: 400 }}>
      <HvProgressBar
        value={value}
        status={status as HvProgressBarStatus}
        labelProps={{
          "aria-label": ariaLabel,
          "aria-busy": false,
          "aria-live": ariaLive,
        }}
      />
      <div style={{ marginTop: 10 }}>
        <HvButton onClick={start}>Start</HvButton>
        <HvButton style={{ marginLeft: 10 }} onClick={reset}>
          Reset
        </HvButton>
      </div>
    </div>
  );
};
