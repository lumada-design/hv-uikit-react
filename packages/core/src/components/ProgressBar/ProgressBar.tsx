import clsx from "clsx";
import clamp from "lodash/clamp";
import { HvBaseProps } from "../../types";
import {
  StyledProgressBar,
  StyledProgressBarContainer,
  StyledProgressContainer,
  StyledRoot,
  StyledValue,
} from "./ProgressBar.styles";
import progressBarClasses, { HvProgressBarClasses } from "./progressBarClasses";

export type HvProgressBarStatus = "inProgress" | "completed" | "error";

/**
 * ProgressBar provides feedback about a process that is taking place in the application.
 */
export type HvProgressBarProps = HvBaseProps & {
  /** The value of the progress bar. */
  value: number;
  /**
   * The status of the progress bar.
   *
   * inProgress is black, error is red and completed is green.
   *
   * When uncontrolled and unspecified it will default to "inProgress".
   */
  status?: HvProgressBarStatus;
  /** Aria Properties passed on to the progress bar. */
  labelProps?: object;
  /** A Jss Object used to override or extend the styles applied to the empty state component. */
  classes?: HvProgressBarClasses;
};

export const HvProgressBar = (props: HvProgressBarProps) => {
  const {
    className,
    classes,
    value = 0,
    status = "inProgress",
    labelProps,
    ...others
  } = props;

  const clampedValue = clamp(value, 0, 100);

  return (
    <StyledRoot
      className={clsx(
        className,
        progressBarClasses.root,
        classes?.root,
        progressBarClasses.progress,
        classes?.progress
      )}
      role="progressbar"
      aria-valuemin={0}
      aria-valuemax={100}
      aria-valuenow={clampedValue}
      {...others}
    >
      <StyledProgressContainer>
        <StyledValue
          className={clsx(
            progressBarClasses.progressBarLabel,
            classes?.progressBarLabel
          )}
          variant="caption2"
          style={{ width: `${clampedValue}%` }}
          {...labelProps}
        >
          {`${clampedValue}%`}
        </StyledValue>
        <StyledProgressBarContainer>
          <StyledProgressBar
            style={{ width: `${clampedValue}%` }}
            className={clsx(
              progressBarClasses.progressBar,
              classes?.progressBar
            )}
            $status={status}
          />
        </StyledProgressBarContainer>
      </StyledProgressContainer>
    </StyledRoot>
  );
};
