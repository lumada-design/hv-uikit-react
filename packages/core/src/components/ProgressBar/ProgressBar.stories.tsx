import { Meta, StoryObj } from "@storybook/react";
import {
  HvTypography,
  HvProgressBar,
  HvProgressBarProps,
} from "@hitachivantara/uikit-react-core";
import { ProgressBarSimulator } from "./ProgressBarSimulator";

const meta: Meta<typeof HvProgressBar> = {
  title: "Components/Loading/Progress Bar",
  component: HvProgressBar,
  decorators: [(Story) => <div style={{ padding: 20 }}>{Story()}</div>],
};
export default meta;

export const Main: StoryObj<HvProgressBarProps> = {
  args: {
    value: 0,
    hideLabel: false,
  },
  argTypes: {
    classes: { control: { disable: true } },
    labelProps: { control: { disable: true } },
    value: { control: { type: "range" } },
  },
  render: (args) => {
    return <HvProgressBar aria-label="Status" {...args} />;
  },
};

export const Progressive: StoryObj<HvProgressBarProps> = {
  args: {
    value: 0,
  },
  argTypes: {
    classes: { control: { disable: true } },
    value: { control: { type: "range" } },
  },
  parameters: {
    eyes: { include: false },
  },
  render: () => {
    return (
      <>
        <div
          style={{
            display: "flex",
            flexDirection: "column",
            justifyContent: "space-around",
            alignItems: "center",
            width: 400,
            margin: "auto",
            marginTop: 20,
            marginBottom: 40,
          }}
        >
          <HvTypography variant="label">Success</HvTypography>
          <ProgressBarSimulator
            inc={(v) => v + 10}
            ariaLabel="Determinate Progress Bar"
            ariaLive="assertive"
          />
        </div>
        <div
          style={{
            display: "flex",
            flexDirection: "column",
            justifyContent: "space-around",
            alignItems: "center",
            width: 400,
            margin: "auto",
          }}
        >
          <HvTypography variant="label">Error</HvTypography>
          <ProgressBarSimulator
            inc={(v) => v + 5}
            error={30}
            ariaLabel="Determinate Progress Bar"
          />
        </div>
      </>
    );
  },
};

export const Determinate: StoryObj<HvProgressBarProps> = {
  args: {
    value: 0,
  },
  argTypes: {
    classes: { control: { disable: true } },
    value: { control: { type: "range" } },
  },
  render: () => {
    return (
      <>
        <div
          style={{
            display: "flex",
            flexDirection: "column",
            justifyContent: "space-around",
            alignItems: "center",
            width: 400,
            margin: "auto",
            marginTop: 20,
            marginBottom: 40,
          }}
        >
          <HvTypography variant="label">Success</HvTypography>
          <HvProgressBar
            value={100}
            status="completed"
            aria-label="Example Determined Progress Bar"
          />
        </div>
        <div
          style={{
            display: "flex",
            flexDirection: "column",
            justifyContent: "space-around",
            alignItems: "center",
            width: 400,
            margin: "auto",
            marginTop: 20,
            marginBottom: 40,
          }}
        >
          <HvTypography variant="label">Loading</HvTypography>
          <HvProgressBar
            value={40}
            aria-label="Example Determined Loading Progress Bar"
          />
        </div>
        <div
          style={{
            display: "flex",
            flexDirection: "column",
            justifyContent: "space-around",
            alignItems: "center",
            width: 400,
            margin: "auto",
          }}
        >
          <HvTypography variant="label">Error</HvTypography>
          <HvProgressBar
            value={30}
            status="error"
            aria-label="Example Determined Error Progress Bar"
          />
        </div>
      </>
    );
  },
};
