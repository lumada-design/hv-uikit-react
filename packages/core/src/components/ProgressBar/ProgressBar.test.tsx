import { describe, it } from "vitest";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { HvProgressBar } from "./ProgressBar";
import { ProgressBarSimulator } from "./ProgressBarSimulator";

describe(
  "ProgressBar",
  () => {
    it("should to be defined", () => {
      const { container } = render(<HvProgressBar value={0} />);
      expect(container).toBeDefined();
    });

    it("should render correctly", () => {
      const { container } = render(<HvProgressBar value={0} />);
      expect(container).toMatchSnapshot();
    });

    describe("Static Determinate Value Tests", () => {
      it("completed progress bar", async () => {
        const { container, findByRole, findByText, getByLabelText } = render(
          <HvProgressBar
            value={100}
            labelProps={{
              "aria-label": "My Aria Label",
              "aria-busy": false,
              "aria-live": "polite",
            }}
          />
        );

        const progressBar = await findByRole("progressbar");
        await findByText("100%");
        expect(progressBar).toHaveAttribute("aria-valuenow", "100");

        const bar = getByLabelText("My Aria Label");
        expect(bar).toBeInTheDocument();

        expect(container).toBeInTheDocument();
      });

      it("zero progress bar", async () => {
        const { container, findByRole, findByText, getByLabelText } = render(
          <HvProgressBar
            value={0}
            labelProps={{
              "aria-label": "My Aria Label",
              "aria-busy": false,
              "aria-live": "polite",
            }}
          />
        );
        const progressBar = await findByRole("progressbar");
        await findByText("0%");
        expect(progressBar).toHaveAttribute("aria-valuenow", "0");

        const bar = getByLabelText("My Aria Label");
        expect(bar).toBeInTheDocument();

        expect(container).toBeInTheDocument();
      });

      it("middle progress bar", async () => {
        const { container, findByRole, findByText, getByLabelText } = render(
          <HvProgressBar
            value={50}
            labelProps={{
              "aria-label": "My Aria Label",
              "aria-busy": false,
              "aria-live": "polite",
            }}
          />
        );
        const progressBar = await findByRole("progressbar");
        await findByText("50%");
        expect(progressBar).toHaveAttribute("aria-valuenow", "50");

        const bar = getByLabelText("My Aria Label");
        expect(bar).toBeInTheDocument();

        expect(container).toBeInTheDocument();
      });

      it("out of bounds progress bar (up)", async () => {
        const { container, findByRole, findByText, getByLabelText } = render(
          <HvProgressBar
            value={200}
            labelProps={{
              "aria-label": "My Aria Label",
              "aria-busy": false,
              "aria-live": "polite",
            }}
          />
        );
        const progressBar = await findByRole("progressbar");
        await findByText("100%");
        expect(progressBar).toHaveAttribute("aria-valuenow", "100");

        const bar = getByLabelText("My Aria Label");
        expect(bar).toBeInTheDocument();

        expect(container).toBeInTheDocument();
      });

      it("out of bounds progress bar (down)", async () => {
        const { container, findByRole, findByText, getByLabelText } = render(
          <HvProgressBar
            value={-200}
            labelProps={{
              "aria-label": "My Aria Label",
              "aria-busy": false,
              "aria-live": "polite",
            }}
          />
        );
        const progressBar = await findByRole("progressbar");
        await findByText("0%");
        expect(progressBar).toHaveAttribute("aria-valuenow", "0");

        const bar = getByLabelText("My Aria Label");
        expect(bar).toBeInTheDocument();

        expect(container).toBeInTheDocument();
      });

      it("error", async () => {
        const { container, findByRole, findByText, getByLabelText } = render(
          <HvProgressBar
            value={-200}
            labelProps={{
              "aria-label": "My Aria Label",
              "aria-busy": false,
              "aria-live": "polite",
            }}
          />
        );
        const progressBar = await findByRole("progressbar");
        await findByText("0%");
        expect(progressBar).toHaveAttribute("aria-valuenow", "0");

        const bar = getByLabelText("My Aria Label");
        expect(bar).toBeInTheDocument();

        expect(container).toBeInTheDocument();
      });
    });

    describe("Aria Label", () => {
      it("check aria label", async () => {
        const { container, findByRole, getByLabelText } = render(
          <HvProgressBar
            value={100}
            labelProps={{
              "aria-label": "My Aria Label",
              "aria-busy": false,
              "aria-live": "polite",
            }}
          />
        );
        await findByRole("progressbar");

        const bar = getByLabelText("My Aria Label");

        expect(bar).toBeInTheDocument();
        expect(container).toBeInTheDocument();
      });
    });

    describe("Dynamic Determinate Value Tests", () => {
      it("completed progress bar", async () => {
        const { container, findByRole, findByText } = render(
          <ProgressBarSimulator inc={(v) => v + 75} ariaLabel="Progress Bar" />
        );

        const progressBar = await findByRole("progressbar");
        const button = await findByText("Start");
        await findByText("0%");
        expect(progressBar).toHaveAttribute("aria-valuenow", "0");

        userEvent.click(button);
        await findByText("100%");
        expect(progressBar).toHaveAttribute("aria-valuenow", "100");
        expect(container).toBeInTheDocument();
      });

      it("error progress bar", async () => {
        const { container, findByRole, findByText } = render(
          <ProgressBarSimulator
            inc={(v) => v + 15}
            ariaLabel="Progress Bar"
            error={30}
          />
        );

        const progressBar = await findByRole("progressbar");
        const button = await findByText("Start");
        await findByText("0%");

        userEvent.click(button);
        await findByText("30%");
        expect(progressBar).toHaveAttribute("aria-valuenow", "30");
        expect(container).toBeInTheDocument();
      });
    });
  },
  { timeout: 30000 }
);
