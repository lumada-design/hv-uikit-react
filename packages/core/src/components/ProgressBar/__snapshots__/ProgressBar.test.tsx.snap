// Vitest Snapshot v1

exports[`ProgressBar > should render correctly 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="0"
    class="HvProgressBar-root HvProgressBar-progress css-vzi3yb-StyledRoot eafrvc94"
    role="progressbar"
  >
    <div
      class="css-1ezqlcy-StyledProgressContainer eafrvc93"
    >
      <p
        class="HvProgressBar-progressBarLabel eafrvc92 css-oaj17d-getStyledComponent-StyledValue e1te0rfk0"
        style="width: 0%;"
        variant="caption2"
      >
        0%
      </p>
      <div
        class="css-179kxy2-StyledProgressBarContainer eafrvc91"
      >
        <div
          class="HvProgressBar-progressBar css-1q42i5j-StyledProgressBar eafrvc90"
          style="width: 0%;"
        />
      </div>
    </div>
  </div>
</div>
`;
